src/setupWorkerModule.ts

```diff
@@ -0,0 +1,24 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License.
+
+import { HookCallback } from '@azure/functions-worker';
+import { Disposable } from './Disposable';
+import { WorkerChannel } from './WorkerChannel';
+import Module = require('module');
+
+export function setupWorkerModule(channel: WorkerChannel): void {
+    const workerApi = {
+        registerHook: (hookName: string, callback: HookCallback) => channel.registerHook(hookName, callback),
+        Disposable,
+    };
+
+    Module.prototype.require = new Proxy(Module.prototype.require, {
+        apply(target, thisArg, argArray) {
+            if (argArray[0] === '@azure/functions-worker') {
```

Comment:
Those reasons make a lot of sense to me.

Do we have a name in mind for this "fancier" programming model package? I agree that the best name for it would be `@azure/functions`, but then we would still run into the first problem that you described. This might be a very premature discussion, but I think just having some temporary tentative name could help clear out some confusion here.

Let's say the programming model package becomes `@azure/functions` while the types package is `@types/azure-functions`. If the intended audience of `@azure/functions-worker` is mainly the authors of `@azure/functions`, then I'm on board with that name :) My only argument would be that we should also expose these hooks in `@azure/functions`.

Reply:
My preferred name for the "programming model package" at the moment is `@azure/functions`.

> Let's say the programming model package becomes @azure/functions while the types package is @types/azure-functions

We wouldn't need a separate types package for `@azure/functions`. Most npm modules ship the types as a part of their main module and we would do the same. We only need a separate types package for the worker package because it's built-in and not shipped with the user's app.

> My only argument would be that we should also expose these hooks in @azure/functions.

Yes I view this as a requirement for both packages

Another thing I haven't mentioned is that it would be super easy to support _another_ name for the worker package here since it's just an `if` statement. If we decide the worker package needs to be `@azure/functions` down the line, all it would take is `if (argArray[0] === '@azure/functions-worker' || argArray[0] === '@azure/functions') {`

Reply:
Yeah my preferred name would be `@azure/functions` as well. This comment here

> 1. That name is currently used for our types-only package. I think users would be confused if they're using something other than a type (an actual function `registerHook`) from that package all of the sudden

implied to me that we want to reserve `@azure/functions` for types only. But my preference would be what you described anyway!

> Another thing I haven't mentioned is that it would be super easy to support _another_ name for the worker package here since it's just an `if` statement. If we decide the worker package needs to be `@azure/functions` down the line, all it would take is `if (argArray[0] === '@azure/functions-worker' || argArray[0] === '@azure/functions') {`

Great point :)